### PR TITLE
Doctrine PSR container

### PR DESCRIPTION
**Pull request by imonteiro**
It's still need test if Doctrine works if both dasprid/container-interop-doctrine and doctrine/doctrine-orm-module are used.

Please, see https://github.com/juriansluiman/SlmQueueDoctrine/issues/145#issuecomment-474785002.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -34,7 +33,7 @@ Before reading SlmQueueDoctrine documentation, please read [SlmQueue documentati
 
 ### Configuring the connection
 
-You need to register a doctrine connection which SlmQueueDoctrine will use to access the database into the service manager. Here is some more [information](https://github.com/doctrine/DoctrineORMModule#connection-settings).
+You need to register a doctrine connection which SlmQueueDoctrine will use to access the database into the service manager. Here is some [examples](https://github.com/DASPRiD/container-interop-doctrine/tree/master/example).
```
Comment (basz) [inline README.md]: Here *are* some examples

File: composer.json [hunk 2]
```diff
@@ -26,7 +26,7 @@
         "zendframework/zend-servicemanager": "^2.7 || ^3.0",
         "zendframework/zend-stdlib": "^2.2 || ^3.0",
         "slm/queue": "^1.0",
-        "doctrine/doctrine-orm-module":  "~0.7 || ^1.0 || ^2.1",
+        "dasprid/container-interop-doctrine": "^1.1",
         "doctrine/dbal": "^2.5"
     },
     "require-dev": {
```
Comment (basz) [inline composer.json]: perhaps we can use the "suggest" key of composer to suggest using `doctrine/doctrine-orm-module`

File: tests/Asset/OMJob.php [hunk 3]
```diff
@@ -3,10 +3,9 @@
 namespace SlmQueueDoctrineTest\Asset;
 
 use Doctrine\Common\Persistence\ObjectManager;
-use DoctrineModule\Persistence\ObjectManagerAwareInterface;
 use SlmQueue\Job\AbstractJob;
 
-class OMJob extends AbstractJob implements ObjectManagerAwareInterface
```
Comment (basz) [inline tests/Asset/OMJob.php]: This is used in a strategy. I guess we should add a note about this?

https://github.com/juriansluiman/SlmQueueDoctrine/search?q=ObjectManagerAwareInterface

File: composer.json (hunk 2 above)
Comment (imonteiro) [reply composer.json]: `"doctrine/doctrine-orm-module": "if you want use Doctrine ORM Module"` is ok? Or we should set another description?

File: tests/Asset/OMJob.php (hunk 3 above)
Comment (imonteiro) [reply tests/Asset/OMJob.php]: We should create a Interface inside SlmQueueDoctrine namespace or just put in README the information about the functions needed (`public function setObjectManager(ObjectManager $objectManager)` and `public function getObjectManager()`)?

File: composer.json (hunk 2 above)
Comment (basz) [reply composer.json]: Perhaps, ... 

"doctrine/doctrine-orm-module" : "^2.0 to use DoctrineORMModule as dasprid/container-interop-doctrine implementation"

File: tests/Asset/OMJob.php (hunk 3 above)
Comment (basz) [reply tests/Asset/OMJob.php]: A note is fine I think. Most people will use DoctrineORMModule anyway I'm guessing.

Additionally, we should note that DoctrineORMModule is not a direct requirement anymore and people should require it in their root composer.json or a break occurs.

File: composer.json (hunk 2 above)
Comment (imonteiro) [reply composer.json]: Done

File: tests/Asset/OMJob.php (hunk 3 above)
Comment (imonteiro) [reply tests/Asset/OMJob.php]: Done. Please, review the note.

File: README.md (hunk 1 above)
Comment (imonteiro) [reply README.md]: Done

**Comment by imonteiro** [comment]
@basz @roelvanduijnhoven Can you help me with tests?

**Comment by MatthiasKuehneEllerhold** [comment]
What happens if I want to use the queue on a zend-mvc project? Is "dasprid/container-interop-doctrine" needed or even used?

I'd prefer to add "dasprid" and "doctrine-orm-module" both to suggest so the user may decide which integration he wants.

**Comment by imonteiro** [comment]
So far the tests are passing without `--prefer-lowest`.

> I'd prefer to add "dasprid" and "doctrine-orm-module" both to suggest so the user may decide which integration he wants.

It's a good approach. Let's see what @basz and @roelvanduijnhoven think about it.

**Comment by imonteiro** [comment]
About tests... we should test against both modules, right?

**Comment by roelvanduijnhoven** [comment]
> What happens if I want to use the queue on a zend-mvc project? Is "dasprid/container-interop-doctrine" needed or even used?
> 
> I'd prefer to add "dasprid" and "doctrine-orm-module" both to suggest so the user may decide which integration he wants.

I am also curious how this works. As long as we are not sure: it is hard to decide on how this should work...

If in fact ZF needs zend-mvc, and Expressive needs dasprid that would complicate things. Right? We need some configuration option in that case. But that does not sound great either.

So first step I think is to test whether using dasprid's lib will just work for ZF. Without duplicating connection instances and stuff.

**Comment by basz** [comment]
I think our goal here is to offer choice as `dasprid/container-interop-doctrine` is an *alternative* to `doctrine/doctrine-orm-module `. One wouldn't need to use them both at the same time and which one (if any) is preferred is up to the end user.

Meaning all we need to do is *suggest* them both, but require none. We would also need to run tests twice if we choose to support them both. The only hard requirement would be `doctrine/dbal` but *how* the connection is instantiated is up to the user since we only use `$container->get($queueOptions->getConnection())` which must be an instance of `\Doctrine\DBAL\Connection`. 

I think this would justify a BC release, but I feel in most cases existing installations would already have a requirement on `doctrine/doctrine-orm-module` and continue to work fine. All in all it's a nice generalization.

Travis should do something like

- composer install
- composer require dasprid/container-interop-doctrine
- activate dasprid/container-interop-doctrine  specific config
- run tests
- composer remove dasprid/container-interop-doctrine
- composer require doctrine/doctrine-orm-module
- activate doctrine/doctrine-orm-module specific config
- run tests

**Comment by basz** [comment]
Could you rebase against latest master? Merge conflicts

**Comment by imonteiro** [comment]
Rebase: done.

Some tests are still not passing... this seems to be related with some _composer_ incompatibility. I will check, however if you have some tip I will appreciate 👍 

Beside this, I have one doubt. In this [line](https://github.com/juriansluiman/SlmQueueDoctrine/blob/5e7ccc9418dfebec9e50fac74748eb304de36f65/src/Strategy/ClearObjectManagerStrategy.php#L33) line should we remove `$job instanceof ObjectManagerAwareInterface` validation?

**Comment by basz** [comment]
Thanks!

**Comment by imonteiro** [comment]
👍 

Sorry for insisting... In this [line](https://github.com/juriansluiman/SlmQueueDoctrine/blob/5e7ccc9418dfebec9e50fac74748eb304de36f65/src/Strategy/ClearObjectManagerStrategy.php#L33) should we remove `$job instanceof ObjectManagerAwareInterface` validation?

Depending on answer, we should uncomment this [line](https://github.com/juriansluiman/SlmQueueDoctrine/blob/a4be4f5be75204e7c5b2a02448b35a6cc1b86d9b/tests/Strategy/ClearObjectManagerStrategyTest.php#L52) or do something like this `if(getenv('DBAL_PROVIDER') === 'doctrine-orm-module')`

**Comment by basz** [comment]
not at all. I was surprised the tests doesn’t fatally error or fails when we run tests without the interface installed. Now I see you commented that line... :)

I believe $this->markAsSkipped() would be appropriate inside those tests. But rather then checking for the DBAL_PROVIDER we should use the interface_exists method for this.

**Comment by imonteiro** [comment]
I don't see why this Strategy can't be used with _container-interop-doctrine_... I'm wrong?

**Comment by basz** [comment]
> If not, and to avoid BC, we could create and Interface (like DoctrineModule) and write something like:
if (($job instanceof ObjectManagerAwareInterface || $job instanceof NewObjectManagerAwareInterface) && $job->getObjectManager())

Yes, that would work.

@roelvanduijnhoven what do you think about BC. Should we do a minor or major release for this change? Are there any plans to release a major version of slm-queue (perhaps php7+) anytime soon that I should consider?

**Pull request merged**